### DISCUSSION TITLE
Embed TUM-Live in Android Exoplayer

### DISCUSSION THREAD

**Discussion by TimOrtel** [TUM-Dev/gocast#893] [Q&A]
Hello,
for the Artemis Android application I would like to embed a tum live video in an Android Application, preferably using [Exoplayer](https://exoplayer.dev/). 

Let's say I have the following url: https://live.rbg.tum.de/w/Sem/20489
How can I derive a link from that such that I can forward it to Exoplayer? The Exoplayer documentation can be found [here](https://exoplayer.dev/media-items.html).

**Comment by joschahenningsen** [comment]
TUM-Live currently doesn't support streams for third party apps out of the box. One reason is that we are responsible for handling authorization and authentication and making sure videos can only be played by users that are allowed to see a course. That also includes handling changes in a courses metadata (we can't deliver private streams to unauthenticated users even when the course used to be public when the link was generated). 

This is achieved by adding a json web token to our (vod) hls playlists. The resulting playlist looks something like this (use https://jwt.io to decode the key):
```
https://edge.live.rbg.tum.de/vod/Sem_2023_01_16_07_30COMB.mp4/playlist.m3u8?jwt=eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE2NzQyNjA1NjksIlVzZXJJRCI6MCwiUGxheWxpc3QiOiJodHRwczovL2VkZ2UubGl2ZS5yYmcudHVtLmRlL3ZvZC9TZW1fMjAyM18wMV8xNl8wN18zMENPTUIubXA0L3BsYXlsaXN0Lm0zdTgifQ.WWovDXZS37bYFhqGOSIfsgfMVzXD3gXbs_t9eFCj4J56B2gZhlRgNq0-qdg5tgG7k8ePuUfizupc7L_dFcbMuwOyoZ2m6Rny9xYuMDqmgCeXc9Tii1MD9m-Y2W93vjmG0MHCgPYLeTtK0_ANkweRbq3Jncq04ktYkYydcA0_D95iyiR3kEVBSA9xllmAlKfPx02ehAp1-wVwereDaZMj8tYFy3by-XwiHuK7iuunNQqwzc8LzsmVBp0qS2Z_s_QuDB-Dx9aAV0DWn2N9qpD8YuirglQBQ0AcT5gTVLOlXiNoVIjgh_m5gVsqlJvgC6kZ5ZHJlekNSw4IhqBDbk8tPw
```

The jwt is uniquely generated for every user visiting the site and contains:
- an expiry date (currently 7 hours after issuing)
- the stream the token allows to play
- the userUD (or 0 if not logged in) of the user playing the stream.

We also consider adding the users IP address to prevent links from being shared.
The public part of the key used to sign the jwt is found at https://tum.live/jwtPubKey

To implement your feature I could imagine a (gRPC/REST) endpoint that Artemis authenticates against using a token that allows generating a signed hls playlist for public and hidden streams (streams that can only be seen if you know the URL).

Feel free to share your thoughts on how this can be designed or schedule a meeting with me (mailto:firstname.lastname@tum.de)

I want to mention however that TUM-Live is much more than a dropbox for your videos that can be delivered via http. By only embedding a video somewhere users lose many features (e.g. the split view player: https://live.rbg.tum.de/w/Sem/20488/SPLIT, automatic silence skipping, bookmarks, previews when scrubbing the seek bar, chat, issue reporting and many more). What you want to achieve can be done by statically delivering a .mp4 through Artemis. Surely this wouldn't be an issue using lecture attachments (ping @krusche).

Looking forward to hear your thoughts on this! 😊

**Reply by krusche** [reply]
We could store videos on Artemis, however this would mean we would re-build parts of TUM.Live, which I would like to avoid. In addition, we currently do not have enough storage to support this and we might run into bandwidth problems. This is simply not our core business.

In addition, we could not use this for livestreams.

I think it would be really cool if Android and iOS users could watch TUM.Live videos in their Artemis app in the native video player.

**Reply by TimOrtel** [reply]
I think one possible approach for this could be a tum live android and iOS library that can be used to embed a tum live video while maintaining features such as the split view and allowing other mobile applications to use tum live in a standardized way, too.

**Reply by joschahenningsen** [reply]
I think we can add an endpoint 
- `https://live.rbg.tum.de/w/Sem/20488/COMB/playlist`
- `https://live.rbg.tum.de/w/Sem/20488/CAM/playlist`
- `https://live.rbg.tum.de/w/Sem/20488/PRES/playlist`

Which redirects to a signed playlist which can be used in Exoplayer (e.g. `302 Found -> https://edge.live.rbg.tum.de/vod/Sem_2023_01_16_07_30COMB.mp4/playlist.m3u8?jwt=eyabc.eydef.xyz`)
This would (like the current embeds in Artemis) be limited to public and hidden streams. Also lecturers won't have access to viewer statistics from app users without significant changes on how we gather statistics.
